miner/worker.go

```diff
@@ -171,32 +175,28 @@ func (self *worker) setExtra(extra []byte) {
 }
 
 func (self *worker) pending() (*types.Block, *state.StateDB) {
-	self.currentMu.Lock()
-	defer self.currentMu.Unlock()
-
 	if atomic.LoadInt32(&self.mining) == 0 {
-		return types.NewBlock(
-			self.current.header,
-			self.current.txs,
-			nil,
-			self.current.receipts,
-		), self.current.state.Copy()
+		// return a snapshot to avoid contention on currentMu mutex
+		self.snapshotMu.RLock()
+		defer self.snapshotMu.RUnlock()
+		return self.snapshotBlock, self.snapshotState.Copy()
```

Comment:
NOTE: Here we are returning the *same* snapshotBlock to all consumers, rather than a copy each time.  That seemed ok to me in my testing, but we could be more conservative and use `types.NewBlock(...)` here to make another copy of the `snapshotBlock`.  Same applies for `pendingBlock()` below.

Reply:
A block is immutable AFAIK, so it should be safe to return the same thing for everyone.